src/transaction.rs

```diff
@@ -629,6 +631,120 @@ impl Transaction {
         Wtxid::from_engine(enc)
     }
 
+    /// Encodes the signing data from which a signature hash for a given input index with a given
+    /// sighash flag can be computed.  To actually produce a scriptSig, this hash needs to be run
+    /// through an ECDSA signer, the SigHashType appended to the resulting sig, and a script
+    /// written around this, but this is the general (and hard) part.
+    ///
+    /// *Warning* This does NOT attempt to support OP_CODESEPARATOR. In general this would require
+    /// evaluating `script_pubkey` to determine which separators get evaluated and which don't,
+    /// which we don't have the information to determine.
+    ///
+    /// # Panics Panics if `input_index` is greater than or equal to `self.input.len()`
+    ///
+    pub fn encode_signing_data_to<Write: io::Write>(
+        &self,
+        mut writer: Write,
+        input_index: usize,
+        script_pubkey: &Script,
+        sighash_type: SigHashType,
+    ) -> Result<(), encode::Error> {
+        assert!(input_index < self.input.len());  // Panic on OOB
```

Comment:
This function is already fallible. Why do we need to panic here?

Reply:
I am worried that, for example, in the case of a multi-party protocol, either the transaction or the index is coming from another party and hence we cannot always guarantee this pre-condition. Having this panic here would mean that we need to explicitly guard against this outside of this function to avoid other parties being able to panic our code.

Reply:
The main function which parties should be using `signature_hash()` is not fallible and we are using `expect` over there when we call this function inside it. But you are right, there is no reason this particular function should panic.